### PR TITLE
Issue #7529: Add ContainerState to BrowserState

### LINKED ISSUE #7529: Add ContainerState to BrowserState

### PR DESCRIPTION
**Pull request by gabrielluong**
- Adds initial container state, actions and reducers to BrowserState
- Refactor `feature/containers/Container.kt` to `browser/state/state/ContainerState.kt`

---

Fixes #7529

### Pull Request checklist

- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### CODE REVIEW

File: components/browser/state/src/main/java/mozilla/components/browser/state/state/ContainerState.kt [hunk 1]
```diff
@@ -2,7 +2,22 @@
...
+ *
+ * @property contextId The session context ID also known as cookie store ID for the container.
+ * @property name Name of the container.
+ * @property color The color for the container. This can be shown in tabs belonging to this container.
+ * @property icon The name of an icon for the container.
+ */
+data class ContainerState(
+    override val contextId: String,
+    override val name: String,
+    override val color: String,
```
Comment (pocmo) [inline components/browser/state/src/main/java/mozilla/components/browser/state/state/ContainerState.kt:18]: I wonder why `Color` is a `String`? Wouldn't it make more sense to have an `Int` here that we can use as a color in UI code?

File: components/browser/state/src/main/java/mozilla/components/browser/state/state/ContainerState.kt [hunk 2]
```diff
@@ -2,7 +2,22 @@
...
+    override val icon: String
```
Comment (pocmo) [inline components/browser/state/src/main/java/mozilla/components/browser/state/state/ContainerState.kt:19]: I'm also wondering about the type here, what does "name of an icon" actually mean? How can I display it? Could this be a Bitmap? Or a Drawable? Or a file path? How do I currently turn this name into something I can display?

File: components/browser/state/src/main/java/mozilla/components/browser/state/state/ContainerState.kt [hunk 3]
```diff
@@ -2,7 +2,22 @@
...
+) : Container
```
Comment (pocmo) [inline components/browser/state/src/main/java/mozilla/components/browser/state/state/ContainerState.kt:20]: Where is this interface coming from? 🤔 Is it needed?

File: components/browser/state/src/main/java/mozilla/components/browser/state/state/BrowserState.kt [hunk 4]
```diff
@@ -13,7 +13,8 @@ import mozilla.components.lib.state.State
  * @property tabs the list of open tabs, defaults to an empty list.
  * @property selectedTabId the ID of the currently selected (active) tab.
  * @property customTabs the list of custom tabs, defaults to an empty list.
- * @property extensions A map of extension ids and web extensions of all installed web extensions.
+ * @property containers A map of session context IDs and their respective container [ContainerState].
```
Comment (NotWoods) [inline components/browser/state/src/main/java/mozilla/components/browser/state/state/BrowserState.kt:16]: The use of session here makes me think of session IDs. Maybe just say "A map of context IDs" or "A map of [SessionState.contextId]"?

Comment (gabrielluong) [reply components/browser/state/src/main/java/mozilla/components/browser/state/state/BrowserState.kt:16]: Thanks, this is a good change. Will have this resolved in my next revision.

File: components/browser/state/src/main/java/mozilla/components/browser/state/state/ContainerState.kt (hunk 1 above)
Comment (gabrielluong) [reply components/browser/state/src/main/java/mozilla/components/browser/state/state/ContainerState.kt:18]: Changed to `@ColorInt val color: Int,` in my next revision.

File: components/browser/state/src/main/java/mozilla/components/browser/state/state/ContainerState.kt (hunk 2 above)
Comment (gabrielluong) [reply components/browser/state/src/main/java/mozilla/components/browser/state/state/ContainerState.kt:19]: Changed to Drawable

File: components/browser/state/src/main/java/mozilla/components/browser/state/state/ContainerState.kt [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+ * @property name Name of the container.
+ * @property color The color for the container. This can be shown in tabs belonging to this container.
+ * @property icon The icon for the container.
+ */
+data class ContainerState(
+    val contextId: String,
+    val name: String,
+    val color: Color,
+    val icon: Icon
+) {
```
Comment (gabrielluong) [inline components/browser/state/src/main/java/mozilla/components/browser/state/state/ContainerState.kt:20]: Curious if this should be a Parcelable from comparing it to `SitePermissions`?

Comment (pocmo) [reply components/browser/state/src/main/java/mozilla/components/browser/state/state/ContainerState.kt:20]: No, I do not think so. I can't see any need for passing instances around inside a bundle (e.g. from activity to activity). Everything can talk to the store directly. If we need to pass anything around then it's probably the id and then the receiver can lookup the right state from the store with that id.

Comment (gabrielluong) [reply components/browser/state/src/main/java/mozilla/components/browser/state/state/ContainerState.kt:20]: Thank you that is very insightful about Parcelable.

**Comment by gabrielluong** [comment]
> ContainerStorage

Yes, I believe we can make this `internal` since only the middleware will needs this from my prototyping.

**Comment by gabrielluong** [comment]
bors r=pocmo

**Pull request merged**